### ISSUE TITLE
Can't properly decode an URL if it contains slugs names that are also parameter names.

### ISSUE THREAD

**Issue by Joebeazelman** [nopSolutions/nopCommerce#5384]
nopCommerce 4.3 

Please read:
[Link to NopCommerce](https://www.nopcommerce.com/en/boards/topic/90303/receiving-404-error-on-the-root-category-after-upgrade-to-43) 

Using a debugger, I set a breakpoint on Line 96 of Category

```
        public virtual IActionResult Category(int categoryId, CatalogPagingFilteringModel command)
        {
            var category = _categoryService.GetCategoryById(categoryId);
            if (category == null || category.Deleted)
                return InvokeHttp404();
....

```
At first, I thought there was something wrong with my debugger setup because I couldn't hit the breakpoint www.store.com/products. When I clicked on another category, however, it worked. The 404 error wasn't generated in the Category method. 

I changed the slug name to "products-test" under **Catalog > Categories > "Products" > Search engine friendly page name ** and it magically again hit my breakpoint. Very strange!  

I also tried to clear the Slug name, but it automatically generated "products" after the category name and still didn't hit the break point.  Finally, I deleted the Slug in **System > Search engine friendly page names** and the Slug field became blank. Navigating, however, to the page didn't result in a 404 error, but gave me a blank page. I checked my configuration to see if I had any URL redirections at the web server level, but I didn't have any. I also cleared the cache to rid any stale values from the system.

I finally found the cause of the error. The fact "products" gave me a 404, but not other slug names lead me to suspect it may be a parsing issue. I tried "category"and "product" as a slug names and it also gave me 404 errors!

The URL routing in NopCommerce 4.3 cannot handle cases where the Slug is the same as a parameter name!  It appears the URL decoding has changed since 4.2. The best fix would be to either correct the URL decoding or add validation logic on all user entered slug names.

**Comment by RomanovM** [comment]
Seems that already fixed (see #4793) in the upcoming 4.40.
But we really should double-check all used routes and add them to the reserved names.

**Comment by RomanovM** [comment]
Actually there are no such problems in the develop branch, since we slightly improve the routing infrastructure and now similar slugs (e.g. "products") work correctly as pattern for generic routes.

We just should update the appropriate settings.
